### DISCUSSION TITLE
Max font size of titles

### DISCUSSION THREAD

**Discussion by ichnilatis-gr** [koreader/koreader#7278] [Q&A]
Could there be an option to set a maximum font size for titles in the file browser, please? Sometimes are too large.

Thank you!

**Comment by NiLuJe** [comment]
There... kinda already is, I think?

(This might become clearer once @poire-z's reorganization of those settings land ;)).

**Reply by poire-z** [reply]
I was thinking this would  be more about one of the CoverBrowser Mosaic or detailed list mode - but as there's not much detail given (which mode, Files per page, a screenshot may be), and given this is asked in the muddy ugly "GitHub Discussions" area that I am boycotting, I won't ask more questions (pfff, I just did :/).

**Reply by ichnilatis-gr** [reply]
There is this option only for the "classic" mode. I'm referring to the detailed list. When a title is short, the letters are too large. Could it be fixed?

**Reply by poire-z** [reply]
> "too large", "fixed"

This is a question of taste, and we can't see your taste (or what you don't taste) as you're not giving us much information. How many "detailed list items per page", and please post a screenshot. This may also depend on your device Screen dpi.

There is already some hardcoded max font sizes for them (that we use as a start, and we decrease it if needed to fit longer text):
https://github.com/koreader/koreader/blob/2348cdb0660791b08f1bc76cf17d01b7368b3735/plugins/coverbrowser.koplugin/listmenu.lua#L500-L503
These were chosen according to our/my taste :) for a few tested values of items per page, and it felt ok.
So, either your tastes are really different, or you have some kind of configuration combinations we haven't tested and somehow screw the estimation. A screenshot will help guessing.

Anyway, may be you can limit the font size by increasing Display mode > ... settings > Detailed list items per pages.
Or tweak the above value to see what you'd be fine with (but I guess we won't change them, as I can't really see a problem with them).

**Reply by ichnilatis-gr** [reply]
Τhank you for your instructions.
For me, the font size of the following titles is too large.
![FileManager_2021-02-13_183855](https://user-images.githubusercontent.com/76203123/107966026-f7abe980-6fb3-11eb-8145-6402f489bf42.png)


So, I set it smaller.
![FileManager_2021-02-15_124244](https://user-images.githubusercontent.com/76203123/107966055-ff6b8e00-6fb3-11eb-88af-1a915ae2f3a4.png)


Could you tell me, please, where can I set the max size of the folders' names?
Also, if I copy these lines in a new file with the name "listmenu.persistent.lua", my settings will remain after the next update?

Thank you.

**Reply by poire-z** [reply]
There was no need to crop and resize your screenshots :/ Now, I have to ask again: **how many items per page ? Really only 5 ? What's your device and screen resolution, what's your DPI ?** (the top title and bottom arrows would have hinted about that...)

**What values did you use instead of 20 and 18 ?**

> where can I set the max size of the folders' names?

https://github.com/koreader/koreader/blob/2348cdb0660791b08f1bc76cf17d01b7368b3735/plugins/coverbrowser.koplugin/listmenu.lua#L233-L235

And everything in that file that use something like `_fontSize(20)` set the font size or max starting font size.

> Also, if I copy these lines in a new file with the name "listmenu.persistent.lua", my settings will remain after the next update?

Nope, the persistent.lua only works for defaults.lua > defaults.persistent.lua.
So, you have to edit these files each time.

But I agree that with 5 items per page, things are getting a bit large :) So, this might need a bit of rethinking.
(I must say I didn't really thought one would use such a low number as 5... I guess it's that you want to see more of the covers, like Mosaic but with title and authors in plain text.)

**Reply by ichnilatis-gr** [reply]
Forgive me for my incomplete answer. I've uploaded again some screenshots in real size ([see above](https://github.com/koreader/koreader/discussions/7278#discussioncomment-370196)). I've set it **6 item per page**. I've chosen it in order for the cover images to be more visible...
I have a **Pocketbook Touch Lux 5** with **212 ppi**.
I've set the values **14** for titles and **10** for authors.
However, it would be easier if there was such an option of setting these numbers.

**Reply by pazos** [reply]
> this is asked in the muddy ugly "GitHub Discussions" area that I am boycotting, I won't ask more questions (pfff, I just did :/).

@poire-z: there's a more efective way of boycotting discussions: disable them on settings. :)

**Comment by poire-z** [comment]
(Follow up to the nested/hidden discussion above - GitHub Discussions: I hate you!)

Or may be we can have something like that:
```diff
--- a/plugins/coverbrowser.koplugin/listmenu.lua
+++ b/plugins/coverbrowser.koplugin/listmenu.lua
@@ -191,15 +191,19 @@ function ListMenuItem:update()
         h = self.height - 2 * self.underline_h
     }

-    local function _fontSize(nominal)
+    local function _fontSize(nominal, no_cap)
         -- nominal font size is based on 64px ListMenuItem height
         -- keep ratio of font size to item height
         local font_size = nominal * dimen.h / 64 / scale_by_size
+        if not no_cap and font_size > 25 then
+            logger.warn("capping to 25", font_size)
+            font_size = 25
+        end
         return math.floor(font_size)
     end
     -- Will speed up a bit if we don't do all font sizes when
     -- looking for one that make text fit
-    local fontsize_dec_step = math.ceil(_fontSize(100) / 100)
+    local fontsize_dec_step = math.ceil(_fontSize(100, true) / 100)

     -- We'll draw a border around cover images, it may not be
     -- needed with some covers, but it's nicer when cover is
```
Or better, try to tweak that formula so it slows the increase of font size when dimen.h is really larger than 64 - instead of doing it linearly. But I suck at maths :)

**Reply by ichnilatis-gr** [reply]
Thank you for overcoming your scruples and using "GitHub Discussions" in order to help me.
However, I'm not good in programming and I can not understand exactly your last piece of advice.

Thank you again.

**Reply by poire-z** [reply]
That was just me throwing out ideas on how to solve this. You're not required to catch these ideas :)
But if some other people have other ideas, or are good at maths :), they are welcome.
I'll probably have a look, may be with something like _fontSize(20, 25) to set the nominal font size, and the max allowed font size, after the nominal is corrected to adjust to the per_page value.

**Reply by ichnilatis-gr** [reply]
Should I open a new issue with this request? Or does the Discussion automatically become a new issue?

**Reply by Frenzie** [reply]
Issues have "convert to discussion" but discussions only have "delete discussion". :thinking:

**Reply by ichnilatis-gr** [reply]
:smile:
So, I think that I should open a new issue. Maybe somebody could help. What do you think?

**Comment by poire-z** [comment]
I've made a tweak to provide a max font size to each thing that gets a nominal font size that would be scaled depending on the number of items per page.
```diff
--- a/plugins/coverbrowser.koplugin/listmenu.lua
+++ b/plugins/coverbrowser.koplugin/listmenu.lua
@@ -50,3 +50,2 @@ local corner_mark
 local scale_by_size = Screen:scaleBySize(1000000) / 1000000
-local max_fontsize_fileinfo

@@ -193,7 +192,12 @@ function ListMenuItem:update()

-    local function _fontSize(nominal)
-        -- nominal font size is based on 64px ListMenuItem height
-        -- keep ratio of font size to item height
-        local font_size = nominal * dimen.h / 64 / scale_by_size
-        return math.floor(font_size)
+    local function _fontSize(nominal, max)
+        -- The nominal font size is based on 64px ListMenuItem height.
+        -- Keep ratio of font size to item height
+        local font_size = math.floor(nominal * dimen.h / 64 / scale_by_size)
+        -- But limit it to the provided max, to avoid huge font size when
+        -- only 4-6 items per page
+        if max and font_size >= max then
+            return max
+        end
+        return font_size
     end
@@ -228,3 +232,3 @@ function ListMenuItem:update()
             text = self.mandatory_func and self.mandatory_func() or self.mandatory,
-            face = Font:getFace("infont", math.min(max_fontsize_fileinfo, _fontSize(15))),
+            face = Font:getFace("infont", _fontSize(14, 16)),
         }
@@ -234,3 +238,3 @@ function ListMenuItem:update()
             text = BD.directory(self.text),
-            face = Font:getFace("cfont", _fontSize(20)),
+            face = Font:getFace("cfont", _fontSize(20, 24)),
             width = wleft_width,
@@ -441,3 +445,3 @@ function ListMenuItem:update()

-            local fontsize_info = math.min(max_fontsize_fileinfo, _fontSize(14))
+            local fontsize_info = _fontSize(14, 16)

@@ -501,4 +505,4 @@ function ListMenuItem:update()
             local fontname_authors = "cfont"
-            local fontsize_title = _fontSize(20)
-            local fontsize_authors = _fontSize(18)
+            local fontsize_title = _fontSize(20, 24)
+            local fontsize_authors = _fontSize(18, 22)
             local wtitle, wauthors
@@ -535,4 +539,4 @@ function ListMenuItem:update()
                     -- as we'll fit 3 lines instead of 2, we can avoid some loops by starting from a lower font size
-                    fontsize_title = _fontSize(17)
-                    fontsize_authors = _fontSize(15)
+                    fontsize_title = _fontSize(17, 21)
+                    fontsize_authors = _fontSize(15, 19)
                 elseif authors then
@@ -565,4 +569,4 @@ function ListMenuItem:update()
                         -- as we'll fit 3 lines instead of 2, we can avoid some loops by starting from a lower font size
-                        fontsize_title = _fontSize(17)
-                        fontsize_authors = _fontSize(15)
+                        fontsize_title = _fontSize(17, 21)
+                        fontsize_authors = _fontSize(15, 19)
                     end
@@ -696,3 +700,3 @@ function ListMenuItem:update()
                 local fileinfo_str = self.mandatory_func()
-                local fontsize_info = math.min(max_fontsize_fileinfo, _fontSize(14))
+                local fontsize_info = _fontSize(14, 16)
                 local wfileinfo = TextWidget:new{
@@ -725,3 +729,3 @@ function ListMenuItem:update()
             local text_widget
-            local fontsize_no_bookinfo = _fontSize(18)
+            local fontsize_no_bookinfo = _fontSize(18, 22)
             repeat
@@ -947,5 +951,2 @@ function ListMenu:_recalculateDimen()

-    -- upper limit for file info font to leave enough space for title and author
-    max_fontsize_fileinfo = available_height / scale_by_size / 32
-
     if self.page_recalc_needed then
```

@ichnilatis-gr : here's the modified file that you can drop and replace yours (keep a backup of yours with your change if you don't want to lose them) `plugins/coverbrowser.koplugin/listmenu.lua`
[listmenu.lua.txt](https://github.com/koreader/koreader/files/6002705/listmenu.lua.txt) (remove the .txt extension)

It may not get as small as yours (especially not for the Authors that you made really smaller - I think it should not go smaller than the format/size/% on the right when there is room) - but just see how it looks for you (at your wished per-page setting, but also at others, just to check with me it looks fine) and if, with this, you would still feel the need to have opened this issue :)

**Reply by ichnilatis-gr** [reply]
You are very good! Thank you very much for your kindness! I will test it.
Will it be embedded in a next version?

**Reply by poire-z** [reply]
If it it good enough for you, yes.
But note that it's not what you requested: "an option to set a maximum font size for titles".
There's no option, the code is just trying to do the best thing. You will tell us if it does :)

**Reply by ichnilatis-gr** [reply]
Well, I think the titles seem fine! As for the authors, I just tried to give to them the same appearance that Pocketobook file browser does. Never mind... they are OK.
However, what would you say, if the details at the right side were a liiiiiiittle bit bigger (Perhaps 1pt bigger), as well as the folder names (1 or 2 pt bigger)?
I upload a screenshot for you to see how they look now.
![FileManager_2021-02-18_172959](https://user-images.githubusercontent.com/76203123/108497579-51fac200-72b4-11eb-9d38-3eba26fd21ed.png)

Let me just ask you: An option for this (as it already exists for the highlights and bookmarks e.t.c.) would be something very difficult? So that everyone can make their own choices... Of course –now that I think about it– that means a lot of choice parameters. Which is rather complicated... :thinking:

Also, I would like to ask: Can I keep your file with the next versions? I mean, will there be no changes or improvements to this file (listmenu.lua)?

**Reply by poire-z** [reply]
> Let me just ask you: An option for this (as it already exists for the highlights and bookmarks e.t.c.) would be something very difficult? So that everyone can make their own choices... Of course –now that I think about it– that means a lot of choice parameters. Which is rather complicated... thinking

Indeed, complicated and difficult. You want to decrease "authors", increase "file info" - I did want to increase "authors" and decrease "file info". Settings will be complicated if we want to allow tweaking each one individually - and these tweaking to apply correctly also when "per page" change. Allowing a single "font_size variation from nominal" (+1 / +2 / -1 / -2) to change them all consistently from the nominal value will make both you and I not happy :) as we wanted some things to increase and other to decrease - so dropping this idea :)
So, easier to just try to do what's "good enough for most" - and people can live with the little deviations from what they would consider "best".

> However, what would you say, if the details at the right side were a liiiiiiittle bit bigger (Perhaps 1pt bigger), as well as the folder names (1 or 2 pt bigger)?

It was originally a bit bigger: `fontsize_info = _fontSize(14, 18)` (max=18). I thought there were a bit too big, taking some valuable space from the title/authors. So, I changed that 18 to 16: `fontsize_info = _fontSize(14, 16)`
BUT please see how it looks with opened books: `51 % of 1135 pages` is quite longer than just `2.5 MB epub`.

You can try changeing it back to 18. Just search for `_fontSize(x, y)` in the file, read around to see what they are about, so you can see what you need tweaking.
Please don't change the x value (nominal value) as it applies to all per_pages - but you can tweak a bit the y value (max), as it will apply only to per_page = 4,5,6, may be 7.


(These values, 14, 16, 18 are a bit abstract: they are not pt, they are scaled depending on your screen size and DPI - but it's consistent with the values we use elsewhere: most titles are 26 or 24, other things 22 or 20, small text 18 or 16. But, for this listmenu.lua, they are additionally scaled depending on the nb of items per page...)

> Also, I would like to ask: Can I keep your file with the next versions? I mean, will there be no changes or improvements to this file (listmenu.lua)?

If we're both fine with some final choices for the values, it will be in next versions with our modifications - so you won't have anything to update - and other users of per_pages=4/5/6/7 will benefit from it (if we both consider it is better than before, I think they will :)

**Reply by ichnilatis-gr** [reply]
Sorry for my delay to reply.
Finally, I will agree with you. I tried the info size 14, 18 and I realized what you said, that it will take valuable space from title/ authors. Also I understood your skeptical about the authors size, that it should not be  smaller than the info size, if the title takes many lines.
So you can keep your tweak, as you made it firstly.

Thank you, once again, for your help, your time and your patience with me... in the discussion that became a long issue... :grinning:

**Reply by poire-z** [reply]
A bit too late :/ I made max=18 in #7324 - which we merged last evening, so it's in today's nightly build.
You'll tell me (in #7324 please :) if that is really bothering, and we'll put it back to 16 (or 17, if this really cause a difference, not sure).

**Reply by ichnilatis-gr** [reply]
It isn't so bothering for me. After all, it was much bigger before and, if it becomes too small, it may bother some people who have vision problems...
I trust you. You know better.
Sorry if I bored you.

**Comment by poire-z** [comment]
Latest bit of code above, with the 16 increased to 18, PR'ed in #7324.
